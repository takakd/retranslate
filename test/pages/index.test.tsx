import 'react'
import Index from '../../pages/index'
import { act, fireEvent, render } from '../testUtils'

const mockTranslateText = jest.fn()

jest.mock('../../api/translator', () => {
  return {
    Translator: jest.fn().mockImplementation(() => {
      return {
        translateText: mockTranslateText,
      }
    }),
    TranslatorOutput: jest.fn(),
  }
})

beforeEach(() => {
  mockTranslateText.mockClear()
})

describe('Index', () => {
  it('matches snapshot', async () => {
    const { asFragment } = render(<Index apiUrl={'https://example.com'} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('select language type', () => {
    const { container } = render(<Index apiUrl={'https://example.com'} />)

    const dropdownItem = container.getElementsByClassName('dropdown-item')

    act(() => {
      // Set source language type to English.
      fireEvent.click(dropdownItem[0])
    })

    const dropdown = container.getElementsByClassName('dropdown-toggle')
    expect(dropdown[0].textContent).toBe('English')
    expect(dropdown[1].textContent).toBe('Japanese')
  })

  it('API', async () => {
    const apiOutput = [
      'translated by aws',
      'translated by google',
      'aws translated by aws',
      'aws translated by google',
      'google translated by aws',
      'google translated by google',
    ]

    mockTranslateText
      .mockResolvedValueOnce({
        // Translate input text.
        aws: apiOutput[0],
        google: apiOutput[1],
      })
      .mockResolvedValueOnce({
        // Retranslate AWS translated text.
        aws: apiOutput[2],
        google: apiOutput[3],
      })
      .mockResolvedValueOnce({
        // Retranslate Google translated text.
        aws: apiOutput[4],
        google: apiOutput[5],
      })

    const { container } = render(<Index apiUrl={'https://example.com'} />)

    const textarea = container.getElementsByTagName('textarea')

    jest.useFakeTimers()

    // Set language.
    const dropdownItem = container.getElementsByClassName('dropdown-item')
    fireEvent.click(dropdownItem[0])

    // Input text to trigger API calling.
    const inputText = 'input value'
    fireEvent.input(textarea[0], {
      target: {
        value: inputText,
      },
    })

    // Wait to complete to call API.
    await act(async () => {
      await jest.runAllTimers()
    })

    // Check
    //    input text
    expect(textarea[0].textContent).toBe(inputText)
    //    API output text
    for (let i = 1, len = textarea.length; i < len; i++) {
      expect(textarea[i].textContent).toBe(apiOutput[i - 1])
    }
  })

  it('API error', async () => {
    const output = 'api error message'
    mockTranslateText.mockRejectedValue(new Error(output))

    const { container } = render(<Index apiUrl={'https://example.com'} />)

    const textarea = container.getElementsByTagName('textarea')

    jest.useFakeTimers()

    // Set language.
    const dropdownItem = container.getElementsByClassName('dropdown-item')
    fireEvent.click(dropdownItem[0])

    // Input text to trigger API calling.
    const inputText = 'input value'
    fireEvent.input(textarea[0], {
      target: {
        value: inputText,
      },
    })

    // Wait to complete to call API.
    await act(async () => {
      await jest.runAllTimers()
    })

    // Check alert message.
    const alert = container.querySelector('.error-alert > span')
    expect(alert.textContent).toBe(`failed: ${output}`)
  })
})
