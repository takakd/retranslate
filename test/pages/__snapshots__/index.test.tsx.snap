// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-146871871 position-relative root"
  >
    <header
      class="jsx-146871871 navbar navbar-dark bg-info fixed-top header py-0 justify-content-start"
    >
      <img
        alt="Retranslate"
        class="jsx-146871871 navbar-logo"
        src="/logo.svg"
      />
      <span
        class="jsx-146871871 navbar-brand ml-3"
      >
        Retranslate
      </span>
      <a
        aria-label="View source on GitHub"
        class="jsx-541412649 position-absolute github-corner"
        href="https://github.com/takakd/retrans-app"
      >
        <svg
          aria-hidden="true"
          class="jsx-541412649"
          height="100%"
          style="fill: #fff; color: rgb(23, 162, 184); border: 0px;"
          viewBox="0 0 250 250"
          width="100%"
        >
          <path
            class="jsx-541412649"
            d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"
          />
          <path
            class="jsx-541412649 octo-arm"
            d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
            fill="currentColor"
            style="transform-origin: 130px 106px;"
          />
          <path
            class="jsx-541412649 octo-body"
            d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
            fill="currentColor"
          />
        </svg>
      </a>
    </header>
    <main
      class="jsx-146871871 main-container"
    >
      <div
        class="jsx-146871871 container-fluid main-container-inner"
      >
        <div
          class="jsx-146871871 row mt-4"
        >
          <div
            class="jsx-146871871 col-12 d-flex align-items-center"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-secondary"
                type="button"
              >
                Language
              </button>
              <div
                aria-labelledby=""
                class="dropdown-menu"
                style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; margin: 0px;"
                x-placement="bottom-start"
              >
                <button
                  class="dropdown-item"
                >
                  English
                </button>
                <button
                  class="dropdown-item"
                >
                  Japanese
                </button>
              </div>
            </div>
            <span
              class="jsx-146871871 px-3"
            >
              to
            </span>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-secondary"
                type="button"
              >
                Language
              </button>
              <div
                aria-labelledby=""
                class="dropdown-menu"
                style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; margin: 0px;"
                x-placement="bottom-start"
              >
                <button
                  class="dropdown-item"
                >
                  English
                </button>
                <button
                  class="dropdown-item"
                >
                  Japanese
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="jsx-146871871 row px-3"
        >
          <div
            class="jsx-146871871 col text-col"
          >
            <div
              class="jsx-511675277 jsx-2841667737 d-flex flex-row align-items-center w-100 text-input"
            >
              <div
                class="jsx-511675277 jsx-2841667737 d-flex flex-column text-input-text-container"
              >
                <div
                  class="jsx-511675277 jsx-2841667737 text-input-header"
                />
                <div
                  class="jsx-511675277 jsx-2841667737 position-relative w-100 text-input-textarea-container"
                >
                  <textarea
                    class="jsx-511675277 jsx-2841667737 h-100 text-input-textarea"
                    placeholder="Type to translate"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="jsx-146871871 col text-col"
          >
            <div
              class="jsx-1918438050 jsx-3453522329 d-flex flex-row align-items-center w-100 text-box"
            >
              <div
                class="jsx-1918438050 jsx-3453522329 icon"
              >
                <svg
                  fill="currentColor"
                  height="20"
                  viewBox="0 0 16 16"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 8a.5.5 0 0 1 .5-.5h11.793l-3.147-3.146a.5.5 0 0 1 .708-.708l4 4a.5.5 0 0 1 0 .708l-4 4a.5.5 0 0 1-.708-.708L13.293 8.5H1.5A.5.5 0 0 1 1 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="jsx-1918438050 jsx-3453522329 d-flex flex-column text-trans-text-container"
              >
                <div
                  class="jsx-1918438050 jsx-3453522329 text-box-header"
                >
                  <img
                    alt="aws"
                    height="20"
                    src="/logo-aws.png"
                    width="28"
                  />
                </div>
                <div
                  class="jsx-1918438050 jsx-3453522329 position-relative w-100 text-trans-textarea-container"
                >
                  <textarea
                    class="jsx-1918438050 jsx-3453522329 h-100 text-trans-textarea"
                  />
                </div>
              </div>
            </div>
            <div
              class="jsx-723032816 jsx-3453522329 d-flex flex-row align-items-center w-100 text-trans-second"
            >
              <div
                class="jsx-723032816 jsx-3453522329 icon"
              >
                <svg
                  fill="currentColor"
                  height="20"
                  viewBox="0 0 16 16"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 8a.5.5 0 0 1 .5-.5h11.793l-3.147-3.146a.5.5 0 0 1 .708-.708l4 4a.5.5 0 0 1 0 .708l-4 4a.5.5 0 0 1-.708-.708L13.293 8.5H1.5A.5.5 0 0 1 1 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="jsx-723032816 jsx-3453522329 d-flex flex-column text-trans-text-container"
              >
                <div
                  class="jsx-723032816 jsx-3453522329 text-trans-second-header"
                >
                  <img
                    alt="google"
                    height="20"
                    src="/logo-google.png"
                    width="20"
                  />
                </div>
                <div
                  class="jsx-723032816 jsx-3453522329 position-relative w-100 text-trans-textarea-container"
                >
                  <textarea
                    class="jsx-723032816 jsx-3453522329 h-100 text-trans-textarea"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="jsx-146871871 col text-col"
          >
            <div
              class="jsx-665952021 jsx-3068828345 d-flex flex-row align-items-center w-100 text-box"
            >
              <div
                class="jsx-665952021 jsx-3068828345 icon"
              >
                <svg
                  fill="currentColor"
                  height="20"
                  viewBox="0 0 16 16"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 8a.5.5 0 0 1 .5-.5h11.793l-3.147-3.146a.5.5 0 0 1 .708-.708l4 4a.5.5 0 0 1 0 .708l-4 4a.5.5 0 0 1-.708-.708L13.293 8.5H1.5A.5.5 0 0 1 1 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="jsx-665952021 jsx-3068828345 d-flex flex-column text-retrans-text-container"
              >
                <div
                  class="jsx-665952021 jsx-3068828345 text-box-header"
                >
                  <img
                    alt="aws"
                    height="20"
                    src="/logo-aws.png"
                    width="28"
                  />
                </div>
                <div
                  class="jsx-665952021 jsx-3068828345 position-relative w-100 text-retrans-textarea-container"
                >
                  <textarea
                    class="jsx-665952021 jsx-3068828345 h-100 text-retrans-textarea"
                  />
                </div>
              </div>
            </div>
            <div
              class="jsx-961098391 jsx-3068828345 d-flex flex-row align-items-center w-100 mt-3"
            >
              <div
                class="jsx-961098391 jsx-3068828345 icon"
              >
                <svg
                  fill="currentColor"
                  height="20"
                  viewBox="0 0 16 16"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 8a.5.5 0 0 1 .5-.5h11.793l-3.147-3.146a.5.5 0 0 1 .708-.708l4 4a.5.5 0 0 1 0 .708l-4 4a.5.5 0 0 1-.708-.708L13.293 8.5H1.5A.5.5 0 0 1 1 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="jsx-961098391 jsx-3068828345 d-flex flex-column text-retrans-text-container"
              >
                <div
                  class="jsx-961098391 jsx-3068828345 mt-3-header"
                >
                  <img
                    alt="google"
                    height="20"
                    src="/logo-google.png"
                    width="20"
                  />
                </div>
                <div
                  class="jsx-961098391 jsx-3068828345 position-relative w-100 text-retrans-textarea-container"
                >
                  <textarea
                    class="jsx-961098391 jsx-3068828345 h-100 text-retrans-textarea"
                  />
                </div>
              </div>
            </div>
            <div
              class="jsx-961098391 jsx-3068828345 d-flex flex-row align-items-center w-100 mt-3"
            >
              <div
                class="jsx-961098391 jsx-3068828345 icon"
              >
                <svg
                  fill="currentColor"
                  height="20"
                  viewBox="0 0 16 16"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 8a.5.5 0 0 1 .5-.5h11.793l-3.147-3.146a.5.5 0 0 1 .708-.708l4 4a.5.5 0 0 1 0 .708l-4 4a.5.5 0 0 1-.708-.708L13.293 8.5H1.5A.5.5 0 0 1 1 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="jsx-961098391 jsx-3068828345 d-flex flex-column text-retrans-text-container"
              >
                <div
                  class="jsx-961098391 jsx-3068828345 mt-3-header"
                >
                  <img
                    alt="aws"
                    height="20"
                    src="/logo-aws.png"
                    width="28"
                  />
                </div>
                <div
                  class="jsx-961098391 jsx-3068828345 position-relative w-100 text-retrans-textarea-container"
                >
                  <textarea
                    class="jsx-961098391 jsx-3068828345 h-100 text-retrans-textarea"
                  />
                </div>
              </div>
            </div>
            <div
              class="jsx-961098391 jsx-3068828345 d-flex flex-row align-items-center w-100 mt-3"
            >
              <div
                class="jsx-961098391 jsx-3068828345 icon"
              >
                <svg
                  fill="currentColor"
                  height="20"
                  viewBox="0 0 16 16"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 8a.5.5 0 0 1 .5-.5h11.793l-3.147-3.146a.5.5 0 0 1 .708-.708l4 4a.5.5 0 0 1 0 .708l-4 4a.5.5 0 0 1-.708-.708L13.293 8.5H1.5A.5.5 0 0 1 1 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="jsx-961098391 jsx-3068828345 d-flex flex-column text-retrans-text-container"
              >
                <div
                  class="jsx-961098391 jsx-3068828345 mt-3-header"
                >
                  <img
                    alt="google"
                    height="20"
                    src="/logo-google.png"
                    width="20"
                  />
                </div>
                <div
                  class="jsx-961098391 jsx-3068828345 position-relative w-100 text-retrans-textarea-container"
                >
                  <textarea
                    class="jsx-961098391 jsx-3068828345 h-100 text-retrans-textarea"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="jsx-146871871 footer mt-4 mb-0 mx-0"
    >
      <div
        class="jsx-146871871 container m-0"
      >
        <p
          class="jsx-146871871 text-muted small"
        >
          © 2021 takakd
        </p>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
